### DISCUSSION TITLE
Best way to emulate NumPy index arrays

### DISCUSSION THREAD

**Discussion by Tudyx** [rust-ndarray/ndarray#1167] [Q&A] [answered]
As stated in the [doc](https://docs.rs/ndarray/latest/ndarray/doc/ndarray_for_numpy_users/index.html#similarities)  `NumPy`  [index arrays](https://numpy.org/doc/stable/user/basics.indexing.html#index-arrays) is not supported yet.
I have a multidimensional array and i want to emulate the result of  `NumPy` index array. For  instance, if my `data` is multi dimensional i want to acces a part of it only by providing a single index, for instance with `data[0]`. It's ok for me to not have the syntactic sugar `[]` of the index `trait`.
Could you please give me some advise about functions in the `ndarray crate`  i could use to try to emulate that?

**Comment by CarlKCarlK** [comment]
Not sure this is exactly on point, but I'm working on supporting fancy indexing within a new Rust project. I'm porting [Bed-Reader](https://github.com/fastlmm/bed-reader/tree/rustybed), our genomics program, from Python to Rust. The program reads 2-D data from a format called PLINK Bed files. The data can be bigger than 1 million individuals (rows) by 1 million DNA locations (columns). When the file is large, users want read just a slice at a time.

The solution I'm creating supports indexing by nothing ("read all"), a number, a negative number (read from end), a vector or array of indexes (including negatives), any Rust range, any ndarray slice, Boolean vectors or arrays. I've put some examples below.

The approach uses a builder pattern. Behind the scenes, it automatically converts all the supported indexing methods into an enum called Index. That gets converted to a vector of indexes when applied to an ndarray of known size.

The code is open source. Let me know if there is interest making this its own crate.
- Carl

*Read individual (samples) from 20 to 30 and every second SNP (variant).*
```
use bed_reader::ReadOptions;
use ndarray::s;

let file_name = "bed_reader/tests/data/some_missing.bed";
let mut bed = Bed::new(file_name)?;
let val = ReadOptions::builder()
    .iid_index(s![..;2])
    .sid_index(20..30)
    .f64()
    .read(&mut bed)?;

assert!(val.dim() == (50, 10));
```
*Read every value in chromosome 5.*

```
let mut bed = Bed::new(file_name)?;
let val = ReadOptions::builder()
    .sid_index(bed.chromosome()?.map(|elem| elem == "5"))
    .f64()
    .read(&mut bed)?;

assert!(val.dim() == (100, 6));
```

```
use ndarray as nd;
use bed_reader::Bed;
use bed_reader::assert_eq_nan;
use bed_reader::ReadOptions;
use ndarray::s;

let file_name = "bed_reader/tests/data/some_missing.bed";
let mut bed = Bed::new(file_name)?;

// Read the individual at index position 3

let val = ReadOptions::builder()
    .iid_index(3)
    .f64()
    .read(&mut bed)?;
assert!(val.dim() == (1, 100));

// Read the individuals at index positions 0, 5, and 1st-from-last.

let val = ReadOptions::builder()
    .iid_index([0, 5, -1])
    .f64()
    .read(&mut bed)?;

assert!(val.dim() == (3, 100));

// Read the individuals at index positions 20 (inclusive) to 30 (exclusive).

let val = ReadOptions::builder()
    .iid_index(20..30)
    .f64()
    .read(&mut bed)?;

assert!(val.dim() == (10, 100));

// Read the individuals at every 2nd index position.

let val = ReadOptions::builder()
    .iid_index(s![..;2])
    .f64()
    .read(&mut bed)?;

assert!(val.dim() == (50, 100));

// Read chromosome 5 of the female individuals.

let female = bed.sex()?.map(|elem| *elem == 2);
let chrom_5 = bed.chromosome()?.map(|elem| elem == "5");
let val = ReadOptions::builder()
    .iid_index(female)
    .sid_index(chrom_5)
    .f64()
    .read(&mut bed)?;

assert!(val.dim() == (50, 6));
```

**Comment by Tudyx** [comment, marked as answer]
Thanks a lot for your proposal! The [Bed-Reader](https://github.com/fastlmm/bed-reader/tree/rustybed) seems to be a very interesting project.
For my problem, is solve it using the `index_axis` method. It's seems to work find for now.
```rust
use ndarray::Axis;
// Emulate the NumPy data[index]
data.index_axis(Axis(0), index)
```
